src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -406,26 +406,66 @@ public function next_tag( $query = null ) {
 		$breadcrumbs  = $query['breadcrumbs'];
 		$match_offset = isset( $query['match_offset'] ) ? (int) $query['match_offset'] : 1;
 
-		$crumb  = end( $breadcrumbs );
-		$target = strtoupper( $crumb );
 		while ( $match_offset > 0 && $this->step() ) {
-			if ( $target !== $this->get_tag() ) {
-				continue;
+			if ( $this->matches_breadcrumbs( $breadcrumbs ) && 0 === --$match_offset ) {
+				return true;
 			}
+		}
 
-			// Look up the stack to see if the breadcrumbs match.
-			foreach ( $this->state->stack_of_open_elements->walk_up() as $node ) {
-				if ( strtoupper( $crumb ) !== $node->node_name ) {
-					break;
-				}
+		return false;
+	}
 
-				$crumb = prev( $breadcrumbs );
-				if ( false === $crumb && 0 === --$match_offset && ! $this->is_tag_closer() ) {
-					return true;
-				}
+	/**
+	 * Indicates if the currently-matched tag matches the given breadcrumbs.
+	 *
+	 * A "*" represents a single tag wildcard, where any tag matches, but not no tags.
+	 *
+	 * At some point this function _may_ support a `**` syntax for matching any number
+	 * of unspecified tags in the breadcrumb stack. This has been intentionally left
+	 * out, however, to keep this function simple and to avoid introducing backtracking,
+	 * which could open up surprising performance breakdowns.
+	 *
+	 * Example:
+	 *
+	 *     $processor = WP_HTML_Processor::createFragment( '<div><span><figure><img></figure></span></div>' );
+	 *     $processor->next_tag( 'img' );
+	 *     true  === $processor->matches_breadcrumbs( array( 'figure', 'img' ) );
+	 *     true  === $processor->matches_breadcrumbs( array( 'span', 'figure', 'img' ) );
+	 *     false === $processor->matches_breadcrumbs( array( 'span', 'img' ) );
+	 *     true  === $processor->matches_breadcrumbs( array( 'span', '*', 'img' ) );
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string[] $breadcrumbs list of HTML tag names representing nested structure.
+	 * @return bool Whether the currently-matched tag is found at the given nested structure.
+	 */
+	public function matches_breadcrumbs( $breadcrumbs ) {
+		if ( ! $this->get_tag() ) {
+			return false;
+		}
+
+		// Everything matches when there are zero constraints.
+		if ( 0 === count( $breadcrumbs ) ) {
+			return true;
+		}
+
+		// Start at the last crumb.
+		$crumb_at = count( $breadcrumbs ) - 1;
```

Comment:
While not likely, it's possible that `$breadcrumbs` may not have indicies that increment uniformly. For example, if someone had these breadcrumbs:

```php
$breadcrumbs = array(
    'html',
    'body',
    'main',
    'img',
);
if ( ! $use_main ) {
    $breadcrumbs = array_diff( $breadcrumbs, array( 'main' ) );
}
```

Then `$breadcrumbs` will now be:

```php
array (
  0 => 'html',
  1 => 'body',
  3 => 'img',
)
```

This would break the iteration below.

So should perhaps this be made more robust by passing `$breadcrumbs through `array_values()` first?

```suggestion
		$breadcrumbs = array_values( $breadcrumbs );
		$crumb_at = count( $breadcrumbs ) - 1;
```

Reply:
interesting. at first I used `end()`/`current()`/`prev()` but changed to the index to avoid mutating the array. I guess I could go back to those.

I'm nervous to start with code that causes unnecessary memory allocation or uncertain runtime performance, even if they are for expectedly small arrays. any thoughts?

Reply:
silly me. we're copying these arrays by passing them into the functions, so I guess we won't risk mutating the passed arrays, and we won't save anything by trying to avoid allocation.

I've restored `end()` and `current()` and `prev()` because of this. I don't think that passing arrays by reference clearly saves on allocations either, but I'm curious to explore that route.